File: src/nuitrack_bridge/types/rgb_frame.cc
#include "nuitrack_bridge/types/rgb_frame.h"
#include "nuitrack-rs/src/nuitrack_bridge/types/rgb_frame.rs.h"
#include "nuitrack/types/RGBFrame.h"
#include "nuitrack/types/Color3.h"
#include "nuitrack/types/Frame.h"

namespace nuitrack_bridge::rgb_frame {

    // --- RGBFrame functions ---
    int32_t getRows(const RGBFrame& frame) {
        return static_cast<int32_t>(frame.getRows());
    }

    int32_t getCols(const RGBFrame& frame) {
        return static_cast<int32_t>(frame.getCols());
    }

    uint64_t getID(const RGBFrame& frame) {
        return frame.getID();
    }

    uint64_t getTimestamp(const RGBFrame& frame) {
        return frame.getTimestamp();
    }

    rust::Slice<const Color3> getData(const RGBFrame& frame) {
        const tdv::nuitrack::Color3* data_ptr = frame.getData();
        size_t num_pixels = static_cast<size_t>(frame.getRows()) * static_cast<size_t>(frame.getCols());
        return rust::Slice<const Color3>{reinterpret_cast<const Color3*>(data_ptr), num_pixels};
    }
}

File: src/nuitrack_bridge/types/rgb_frame.rs
#[cxx::bridge(namespace = "nuitrack_bridge::rgb_frame")]
pub mod ffi {
    
    #[derive{Debug, Clone, Copy, PartialEq}]
    pub struct Color3 {
        pub blue: u8,
        pub green: u8,
        pub red: u8,
    }

    #[namespace = "tdv::nuitrack"]
    unsafe extern "C++" {
        type RGBFrame;
    }

    impl SharedPtr<RGBFrame> {}

    unsafe extern "C++" {
        include!("nuitrack_bridge/types/rgb_frame.h");

        #[cxx_name = "getRows"]
        pub fn rows(frame: &RGBFrame) -> Result<i32>;

        #[cxx_name = "getCols"]
        pub fn cols(frame: &RGBFrame) -> Result<i32>;

        #[cxx_name = "getID"]
        pub fn id(frame: &RGBFrame) -> Result<u64>;

        #[cxx_name = "getTimestamp"]
        pub fn timestamp(frame: &RGBFrame) -> Result<u64>;

        #[cxx_name = "getData"]
        pub fn data(frame: &RGBFrame) -> Result<&[Color3]>;
    }
}

unsafe impl Send for ffi::RGBFrame {}
unsafe impl Sync for ffi::RGBFrame {}